mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java

```diff
@@ -70,138 +68,116 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
         MailboxId mailboxId = previousReIndexingFailure.getMailboxId();
         MessageUid uid = previousReIndexingFailure.getUid();
-        try {
-            return handleMessageReIndexing(mailboxId, uid, reprocessingContext);
-        } catch (MailboxException e) {
-            LOGGER.warn("ReIndexing failed for {} {}", mailboxId, uid, e);
-            reprocessingContext.recordFailureDetailsForMessage(mailboxId, uid);
-            return Task.Result.PARTIAL;
-        }
+
+        return handleMessageReIndexing(mailboxId, uid, reprocessingContext)
+            .onErrorResume(e -> {
+                LOGGER.warn("ReIndexing failed for {} {}", mailboxId, uid, e);
+                reprocessingContext.recordFailureDetailsForMessage(mailboxId, uid);
+                return Mono.just(Result.PARTIAL);
+            });
     }
 
-    Task.Result reIndex(ReprocessingContext reprocessingContext) throws MailboxException {
+    Mono<Result> reIndex(ReprocessingContext reprocessingContext) {
         MailboxSession mailboxSession = mailboxManager.createSystemSession(RE_INDEXER_PERFORMER_USER);
         LOGGER.info("Starting a full reindex");
-        Stream<MailboxId> mailboxIds = mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).list()
-            .stream()
-            .map(Mailbox::getMailboxId);
-
-        try {
-            return reIndex(mailboxIds, reprocessingContext);
-        } finally {
-            LOGGER.info("Full reindex finished");
-        }
+        return mailboxSessionMapperFactory.getMailboxMapper(mailboxSession).list()
+            .concatMap(mailbox -> reIndex(reprocessingContext, mailboxSession, mailbox))
+            .reduce(Task::combine)
+            .switchIfEmpty(Mono.just(Result.COMPLETED))
+            .doFinally(any -> LOGGER.info("Full reindex finished"));
     }
 
-    Task.Result reIndex(Username username, ReprocessingContext reprocessingContext) throws MailboxException {
+    Mono<Result> reIndex(Username username, ReprocessingContext reprocessingContext) throws MailboxException {
         MailboxSession mailboxSession = mailboxManager.createSystemSession(username);
         LOGGER.info("Starting a reindex for user {}", username.asString());
 
-        Stream<MailboxId> mailboxIds = mailboxManager.search(MailboxQuery.privateMailboxesBuilder(mailboxSession).build(), mailboxSession)
-            .stream()
-            .map(MailboxMetaData::getId);
+        MailboxQuery mailboxQuery = MailboxQuery.privateMailboxesBuilder(mailboxSession).build();
 
-        try {
-            return reIndex(mailboxIds, reprocessingContext);
-        } finally {
-            LOGGER.info("User {} reindex finished", username.asString());
-        }
+        return mailboxManager.searchReactive(mailboxQuery, mailboxSession)
+            .map(MailboxMetaData::getId)
+            .concatMap(id -> reIndex(id, reprocessingContext))
```

Comment:
same question: do you need to care about ordering and concurrency?

Reply:
This is a strategy for concurrency control until we get better.

I intend toi start discussions on this, but I see reactification as a preliminary step.

Reply:
Could we use flatMap with concurrency == 1 and prefetch == 1 instead?
Because concatMap **means** something else, it's not a technical detail.